nise/util/__init__.py

```diff
@@ -43,15 +41,22 @@ def load_yaml(objekt):
 
 
 def deepupdate(original, update):
-    """Recursively update a dict.
+    """Recursively update a nested dict/list.
 
-    Subdict's won't be overwritten but also updated.
+    original and update must have the same structure
     """
-    if not isinstance(original, abc.Mapping):
+    if not (isinstance(update, list) or isinstance(update, dict)):
         return update
-    for key, value in update.items():
-        if isinstance(value, abc.Mapping):
+
+    if isinstance(update, dict):
+        for key, value in update.items():
             original[key] = deepupdate(original.get(key, {}), value)
-        else:
-            original[key] = value
+
+    if isinstance(update, list):
+        for idx, item in enumerate(update):
+            if len(original) < idx:
```

Comment:
Shouldn't this always call deepupdate on all items similarly to the dict block above?

Reply:
Lists and dicts are subtley different cases. Unfortunately, lists are more difficult than dicts to do what I'm doing here. If our static files had a syntax that only used dicts and had no lists at all, this would be a much simpler function.

The broader context here is that `original` is the new base template and `update` would be a user-supplied static file. That might be helpful to bear in mind as you're reading the function.

The goal is that we want to overlay, but not overwrite, auto-generated `original` list. We want to keep any auto-generated elements if the user didn't specify some elements of the structure.

So, that's where I look at the length of the two lists.

There's three basic cases:
1. the two lists are the same length.  (i.e. `len(original) == len(update)`)
2. the `original` list has more elements than the `update` list. (i.e. `len(original) > len(update)`)
3. the `original` list has fewer elements than the `update` list. (i.e. `len(original) < len(update)`)

The first case is the easiest. We update each element in the list with whatever the user supplied.

The second case is simliar to the first. We update the first N number of elements in the list and there's M number of auto-generated elements at the end of the list.

The third case means that after we update the existing elements in the `original` list, we then need to `.append()` to the list, rather than access it by index because otherwise Python throws `KeyErrors` for attempting to access a list index that doesn't exist.

Reply:
I think I'm following you now. 
My brain got stuck associating deepupdate with deepcopy and that part of the intent was to be like a deep copy.